### DISCUSSION TITLE
Clean extra newlines from PDF pastes

### DISCUSSION THREAD

**Discussion by xmacex** [SilentVoid13/Templater#644] [Templates Showcase]
While we wait for PDF to fade away into the mist of history, we all struggle with copy-paste from them.

For instance, here is a paste from PDF of Neil Selwyn's [‘*Just playing around with Excel and pivot tables’ - the realities of data-driven schooling*](https://doi.org/10.1080/02671522.2020.1812107) (2020, Research Papers in Education)

> Like any institutional infrastructure, these different elements of 
‘school data’ work to ‘drive and maintain standardization, reflect and embody historical 
concentrations of power and control’ (Dourish and Bell 2007, 416). In one sense, then, all 
the ‘small data’ processes detailed in this paper were clearly reinforcing official priorities 
and institutional agendas within the schools. These were processes that clearly ‘worked’ 
for school authorities. Yet, it is also interesting to note how the minutiae of the data-work 
was affording the individual ‘expert’ teachers left responsible for conducting each analysis 
a considerable degree of ‘tacit’ power. These teachers were making small statistical 
adjustments that nevertheless had sizeable impacts on students’ and teachers’ experiences 
of school. These were relatively low-status teachers within the general hierarchy of their 
schools, yet their analytical decisions afforded them considerable influence over school 
affairs.

Same as a screenshot

![Screenshot_20220523-175943](https://user-images.githubusercontent.com/879300/169860295-ea3357df-c41f-4b84-9d06-65c6f2662262.jpg)

Note the frequent double newlines. I am hoping to find a syntax for cleaning a selected region.

```javascript
<%*
let sel = tp.file.selection()
sel.replace("\n\n", "")
%>
```

was not enough; it removes the selected text altogether.

**Comment by welpdx** [comment]
@xmacex, is this a showcase or do you need help?

### If this is just a showcase, ignore this post
### If you need help: 
Here are my comments:

1) `tp.file.selection()` removes the selection altogether. If you want to display your new `sel` you should add `tR+=sel` at the end. Like so
```js
<%*
var sel = tp.file.selection()
sel.replace("\n\n", "")
tR += sel
console.log(sel)
%>
```
From my experience, copied text from pdfs usually have a combination of `\s`, `\r`, and `\n`. 

Here are some scripts I made for you:

2) Remove whitespace characters from selection:
```js
<%*
let sel = tp.file.selection()
if (sel) {
	sel = sel.replace(/\s*(\n|\r\n)/gm," ").replace(/\r\n/gm," ").replace(/\n/gm," ")
	tR += sel
}
%>
```

3) Remove whitespace characters from clipboard:
```js
<%*
var clip = await tp.system.clipboard();
if (clip) {
	clip = clip.replace(/\s*(\n|\r\n)/gm," ").replace(/\r\n/gm," ").replace(/\n/gm," ")
	tR += clip
}
%>
```
^ So that instead of copy from pdf -> goto obsidian -> paste -> Select -> Run script. You can just copy from pdf -> goto obsidian -> run script. 

![img](https://i.imgur.com/31aCPLT.gif)


Hope this helps

~Welp
